lib/semantics/resolve-names.cc

```diff
@@ -1045,43 +1047,149 @@ class ConstructVisitor : public virtual DeclarationVisitor {
+  static bool NeedsScope(const parser::OpenMPBlockConstruct &);
+
+  bool Pre(const parser::OpenMPBlockConstruct &);
+  void Post(const parser::OpenMPBlockConstruct &);
+  bool Pre(const parser::OmpBeginBlockDirective &x) {
+    AddOmpSourceRange(x.source);
     return true;
   }
-  void Post(const parser::OpenMPBlockConstruct &) { PopScope(); }
-  bool Pre(const parser::OmpBeginBlockDirective &) {
-    ClearDataSharingAttributeObjects();
+  void Post(const parser::OmpBeginBlockDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
+  bool Pre(const parser::OmpEndBlockDirective &x) {
+    AddOmpSourceRange(x.source);
     return true;
   }
+  void Post(const parser::OmpEndBlockDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
 
   bool Pre(const parser::OpenMPLoopConstruct &) {
     PushScope(Scope::Kind::Block, nullptr);
     return true;
   }
   void Post(const parser::OpenMPLoopConstruct &) { PopScope(); }
-  bool Pre(const parser::OmpBeginLoopDirective &) {
-    ClearDataSharingAttributeObjects();
+  bool Pre(const parser::OmpBeginLoopDirective &x) {
+    AddOmpSourceRange(x.source);
+    return true;
+  }
+  void Post(const parser::OmpBeginLoopDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
+  bool Pre(const parser::OmpEndLoopDirective &x) {
+    AddOmpSourceRange(x.source);
+    return true;
+  }
+  void Post(const parser::OmpEndLoopDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
+
+  bool Pre(const parser::OpenMPSectionsConstruct &) {
+    PushScope(Scope::Kind::Block, nullptr);
     return true;
   }
+  void Post(const parser::OpenMPSectionsConstruct &) { PopScope(); }
+  bool Pre(const parser::OmpBeginSectionsDirective &x) {
+    AddOmpSourceRange(x.source);
+    return true;
+  }
+  void Post(const parser::OmpBeginSectionsDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
+  bool Pre(const parser::OmpEndSectionsDirective &x) {
+    AddOmpSourceRange(x.source);
+    return true;
+  }
+  void Post(const parser::OmpEndSectionsDirective &) {
+    messageHandler().set_currStmtSource(std::nullopt);
+  }
+};
+
+bool OmpVisitor::NeedsScope(const parser::OpenMPBlockConstruct &x) {
+  const auto &beginBlockDir{std::get<parser::OmpBeginBlockDirective>(x.t)};
+  const auto &beginDir{std::get<parser::OmpBlockDirective>(beginBlockDir.t)};
+  switch (beginDir.v) {
+  case parser::OmpBlockDirective::Directive::TargetData:
+  case parser::OmpBlockDirective::Directive::Master:
+  case parser::OmpBlockDirective::Directive::Ordered: return false;
+  default: return true;
+  }
+}
 
-  bool Pre(const parser::OpenMPThreadprivate &x) {
+void OmpVisitor::AddOmpSourceRange(const parser::CharBlock &source) {
+  messageHandler().set_currStmtSource(source);
+  for (auto *scope = &currScope(); scope; scope = &scope->parent()) {
+    scope->AddSourceRange(source);
+    if (scope->IsGlobal()) {
+      break;
+    }
+  }
```

Comment:
Maybe this loop should be part of the implementation of `Scope::AddSourceRange` so that it doesn't have to be duplicated here and in `ScopeHandler`.